flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironmentImpl.java

```diff
@@ -0,0 +1,194 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.taskexecutor.partition;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.runtime.deployment.InputGateDeploymentDescriptor;
+import org.apache.flink.runtime.deployment.ResultPartitionDeploymentDescriptor;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.executiongraph.PartitionInfo;
+import org.apache.flink.runtime.io.network.api.writer.ResultPartitionWriter;
+import org.apache.flink.runtime.io.network.partition.PartitionProducerStateProvider;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.shuffle.ShuffleEnvironment;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+/**
+ * Wraps a {@link ShuffleEnvironment} to allow tracking of partitions per job.
+ */
+public class JobAwareShuffleEnvironmentImpl<G extends InputGate> implements JobAwareShuffleEnvironment<NotifyingResultPartitionWriter, G> {
+
+	private static final Consumer<ResultPartitionID> NO_OP_NOTIFIER = partitionId -> {};
+
+	private final ShuffleEnvironment<?, G> backingShuffleEnvironment;
+	private final PartitionTable inProgressPartitionTable = new PartitionTable();
+	private final PartitionTable finishedPartitionTable = new PartitionTable();
+
+	/**	Tracks which jobs are still being monitored, to ensure cleanup in cases where tasks are finishing while
+	 * the jobmanager connection is being terminated. This is a concurrent map since it is modified by both the
+	 * Task (via {@link #notifyPartitionFinished(JobID, ResultPartitionID)}} and
+	 * TaskExecutor (via {@link #releaseAllFinishedPartitionsForJobAndMarkJobInactive(JobID)}) thread. */
+	private final Set<JobID> activeJobs = ConcurrentHashMap.newKeySet();
+
+	public JobAwareShuffleEnvironmentImpl(ShuffleEnvironment<?, G> backingShuffleEnvironment) {
+		this.backingShuffleEnvironment = Preconditions.checkNotNull(backingShuffleEnvironment);
+	}
+
+	@Override
+	public boolean hasPartitionsOccupyingLocalResources(JobID jobId) {
```

Comment:
This method has the similar function with below `getPartitionsOccupyingLocalResources` which might be used for reporting collections of unreleased `partitionIDs` in heartbeat with `ResourceManager`.
Is it possible to remove this method and implement the `getPartitionsOccupyingLocalResources` based on `inProgressPartitionTable` and `finishedPartitionTable`?

Reply:
You'd still need separate methods. This method here is about having partitions for a specific job, while `getPartitionsOccupyingLocalResources()` checks across jobs.

Additionally, we don't track all partitions in the `JobAwareShuffleEnvironment`; only those that are externally managed, so the semantics are a bit different.